webapp/components/utils/UniqueSlugForm.spec.js

```diff
@@ -0,0 +1,80 @@
+import UniqueSlugForm from './UniqueSlugForm'
+import Schema from 'async-validator'
+
+let translate
+let apollo
+let currentUser
+
+beforeEach(() => {
+  translate = jest.fn(() => 'Validation error')
+  apollo = {
+    query: jest.fn().mockResolvedValue({ data: { User: [] } }),
+  }
+  currentUser = null
+})
+
+describe('UniqueSlugForm', () => {
+  let validate = object => {
+    const { formSchema } = UniqueSlugForm({ translate, apollo, currentUser })
+    const validator = new Schema(formSchema)
+    return validator.validate(object, { suppressWarning: true }).catch(({ errors }) => {
+      throw new Error(errors[0].message)
+    })
+  }
+
+  describe('regex', () => {
+    describe('non URL-safe characters, e.g. whitespaces', () => {
+      it('rejects', async () => {
+        await expect(validate({ slug: 'uh oh' })).rejects.toThrow('Validation error')
+      })
+    })
+
+    describe('alphanumeric, hyphens or underscores', () => {
+      it('validates', async () => {
+        await expect(validate({ slug: '_all-right_' })).resolves.toBeUndefined()
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-09-22T11:44:39Z" title="Sunday, September 22nd 2019, 1:44:39 pm +02:00">Sep 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,80 @@
+import UniqueSlugForm from './UniqueSlugForm'
+import Schema from 'async-validator'
+
+let translate
+let apollo
+let currentUser
+
+beforeEach(() => {
+  translate = jest.fn(() => 'Validation error')
+  apollo = {
+    query: jest.fn().mockResolvedValue({ data: { User: [] } }),
+  }
+  currentUser = null
+})
+
+describe('UniqueSlugForm', () => {
+  let validate = object => {
+    const { formSchema } = UniqueSlugForm({ translate, apollo, currentUser })
+    const validator = new Schema(formSchema)
+    return validator.validate(object, { suppressWarning: true }).catch(({ errors }) => {
+      throw new Error(errors[0].message)
+    })
+  }
+
+  describe('regex', () => {
+    describe('non URL-safe characters, e.g. whitespaces', () => {
+      it('rejects', async () => {
+        await expect(validate({ slug: 'uh oh' })).rejects.toThrow('Validation error')
+      })
+    })
+
+    describe('alphanumeric, hyphens or underscores', () => {
+      it('validates', async () => {
+        await expect(validate({ slug: '_all-right_' })).resolves.toBeUndefined()
```


this is confusing for me, it resolves, but to undefined?
can't it return a boolean? or it works like if it fails the validations there is a message to display and if not, there is no message?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-22T21:57:18Z" title="Sunday, September 22nd 2019, 11:57:18 pm +02:00">Sep 22, 2019</time>_

---


Yes, it's the specific implementation of [async-validator](https://github.com/yiminghe/async-validator) where you call a callback on validation error and do nothing otherwise. Thus, no return value.